packages/hyperion-autologging/src/ALSurfaceVisibilityPublisher.ts

```diff
@@ -53,52 +53,110 @@ export type InitOptions = Types.Options<
+        if (index > -1) {
+          values[index] = values[values.length - 1];
+          values.length -= 1;
+          if (values.length === 0) {
+            this.map.delete(key);
+          }
+        }
+      }
+    }
+    [Symbol.iterator]() {
+      return this.map[Symbol.iterator]();
+    }
+  }
+
   // lookup surfaces that are mounted by their root element 
-  const observedRoots = new Map<Element, ALSurfaceData>();
+  const observedRoots = new MapToArray<Element, ALSurfaceData>();
+  const surfaceDataRoots = new MapToArray<ALSurfaceData, Element>();
 
   // We need one observer per threshold
   const observers = new Map<number, IntersectionObserver>();
 
-  channel.addListener('al_surface_mutation_event', event => {
-    __DEV__ && assert(event.surfaceData.getMutationEvent() === event, 'Invalid situation for surface mutation event');
-    switch (event.event) {
-      case 'mount_component': {
-        if (event.capability?.trackVisibilityThreshold) {
-          const observer = getOrCreateObserver(event.capability.trackVisibilityThreshold);
-          const { element } = event;
-          /**
-           * IntersectionObserver cannot track display:content styles because
-           * these elements don't have their own "box".
-           * So, instead we have to focus on the children of the parent element
-           */
-          if (ALSurfaceUtils.isSurfaceWrapper(element)) {
-            for (let el = element.firstElementChild; el; el = el.nextElementSibling) {
-              observer.observe(el);
-              observedRoots.set(el, event.surfaceData);
-            }
-          } else {
-            observer.observe(element);
-            observedRoots.set(element, event.surfaceData);
-          }
-        }
 
-        break;
+  function getNonSurfaceWrapperRoots(element: Element): Element[] {
+    if (ALSurfaceUtils.isSurfaceWrapper(element)) {
+      const subRoots: Element[] = [];
+      for (let el = element.firstElementChild; el; el = el.nextElementSibling) {
+        if (!ALSurfaceUtils.isSurfaceWrapper(el)) {
+          subRoots.push(el);
+        }
       }
-      case 'unmount_component': {
-        if (event.capability?.trackVisibilityThreshold) {
-          const observer = getOrCreateObserver(event.capability.trackVisibilityThreshold);
-          const { element } = event;
-          if (ALSurfaceUtils.isSurfaceWrapper(element)) {
-            for (let el = element.firstElementChild; el; el = el.nextElementSibling) {
-              observer.unobserve(el);
-              observedRoots.delete(el);
-            }
-          } else {
-            observer.unobserve(element);
-            observedRoots.delete(element);
+      if (subRoots.length > 0) {
+        // we found at least one, good enough for us
+        return subRoots;
+      } else {
+        // Could not find non-surface-wrapper root, so need to go deeper
+        for (let el = element.firstElementChild; el; el = el.nextElementSibling) {
```

Comment:
I see so this is the case we discussed needing to account for.

This will continue just 1 level depth,  so it shouldn't be too expensive so long as immediate children isn't large

Reply:
It actually does recurse down, and could be deep, but assumption here is that we will quickly find something other than a surface. 
Also, even a single child along with another surface should work.